### ISSUE TITLE
A reference count handling bug

### ISSUE THREAD

**Issue by sklam** [numba/numba#1526]
https://groups.google.com/a/continuum.io/forum/?utm_medium=email&utm_source=footer#!msg/numba-users/8yKajxuVJ10/CqaZXGA_AwAJ

**Comment by sklam** [comment]
The problem is yet another hidden alias problem I tried to fix in #1518.

The code provided by the user contains a basicblock with the following instructions:

```llvm
  %.1328 = phi { i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] } [ %.1125, %switch.0.2 ], [ %.1126, %switch.1.2 ]
  %extracted.meminfo.36 = extractvalue { i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] } %.1328, 0
  %extracted.shape.36 = extractvalue { i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] } %.1328, 5
  %extracted.strides.36 = extractvalue { i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] } %.1328, 6
  call void @NRT_incref(i8* %extracted.meminfo.36)
  call void @NRT_decref(i8* null)
  %.1328.fca.1.extract = extractvalue { i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] } %.1328, 1
  %.1328.fca.2.extract = extractvalue { i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] } %.1328, 2
  %.1328.fca.3.extract = extractvalue { i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] } %.1328, 3
  %.1328.fca.4.extract = extractvalue { i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] } %.1328, 4
  %.1328.fca.5.0.extract = extractvalue { i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] } %.1328, 5, 0
  %.1328.fca.6.0.extract = extractvalue { i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64] } %.1328, 6, 0
  call void @NRT_decref(i8* %extracted.meminfo.22)
  call void @NRT_decref(i8* %extracted.meminfo.23)
  %.1365 = insertvalue [1 x i64] undef, i64 %.1328.fca.5.0.extract, 0
  %.1366 = extractvalue [1 x i64] %.1365, 0
  %.1368.fca.0.insert = insertvalue [1 x i64] undef, i64 %.1328.fca.6.0.extract, 0
  %.1386 = load i64* %.1328.fca.4.extract
  call void @NRT_decref(i8* %extracted.meminfo.36)
```

The reference count pruning tries to remove the redundant pair of incref and decref `call void @NRT_incref(i8* %extracted.meminfo.36)` and `call void @NRT_decref(i8* %extracted.meminfo.36)`.
However, it does not know that `%extracted.meminfo.22` and/or `%extracted.meminfo.23` are potential alias to `%extracted.meminfo.36` and the following sequence may also access the underlying data:
```llvm
  %.1365 = insertvalue [1 x i64] undef, i64 %.1328.fca.5.0.extract, 0
  %.1366 = extractvalue [1 x i64] %.1365, 0
  %.1368.fca.0.insert = insertvalue [1 x i64] undef, i64 %.1328.fca.6.0.extract, 0
  %.1386 = load i64* %.1328.fca.4.extract
```
After pruning, the above access is loading from data that is removed.

**Comment by sklam** [comment]
#1518 have only fixed problem due to hidden alias related to other incref.  This is hidden alias due to other decref.  A possible solution is to move all the decref to the last location of decref before pruning.
